src/clib/solve_rate_cool_g-cpp.C

```diff
@@ -0,0 +1,1012 @@
+  grackle::impl::View<gr_float***> e(my_fields->internal_energy,
+                                     my_fields->grid_dimension[0],
+                                     my_fields->grid_dimension[1],
+                                     my_fields->grid_dimension[2]);
+
+
+  for (int i = idx_range.i_start; i < idx_range.i_stop; i++) {
+    if (itmask[i] != MASK_FALSE) {
+      // in this case, the chemical network will be evolved with Gauss-Seidel
+
+      // Part 1 of 2: adjust values of dedot and HIdot
+      // ---------------------------------------------
+
+      // Bound from below to prevent numerical errors
+      if (std::fabs(dedot[i]) < tiny8) {
+        dedot[i] = std::fmin(tiny_fortran_val, de(i,j,k));
+      }
+      if (std::fabs(HIdot[i]) < tiny8){
+        HIdot[i] = std::fmin(tiny_fortran_val, HI(i,j,k));
+      }
+
+      // If the net rate is almost perfectly balanced then set
+      //     it to zero (since it is zero to available precision)
+      {
+        double ion_rate = std::fabs(kcr_buf.data[CollisionalRxnLUT::k1][i] *
+                                    de(i,j,k) * HI(i,j,k));
+        double recomb_rate = std::fabs(kcr_buf.data[CollisionalRxnLUT::k2][i] *
+                                       HII(i,j,k) * de(i,j,k));
+        double ratio = (std::fmin(ion_rate, recomb_rate) /
+                        std::fmax(std::fabs(dedot[i]), std::fabs(HIdot[i])));
+        if (ratio > 1.0e6) {
+          dedot[i] = tiny8;
+          HIdot[i] = tiny8;
+        }
+      }
+
+      // If the iteration count is high then take the smaller of
+      //   the calculated dedot and last time step's actual dedot.
+      //   This is intended to get around the problem of a low
+      //   electron or HI fraction which is in equilibrium with high
+      //   individual terms (which all nearly cancel).
+      if (iter > 50)  {
+        dedot[i] = std::fmin(std::fabs(dedot[i]), std::fabs(dedot_prev[i]));
+        HIdot[i] = std::fmin(std::fabs(HIdot[i]), std::fabs(HIdot_prev[i]));
+      }
+
+      // Part 2 of 2: compute minimum rate timestep
+      // ------------------------------------------
+
+      double olddtit = dtit[i];
+      dtit[i] = grackle::impl::fmin(std::fabs(0.1*de(i,j,k)/dedot[i]),
+                                    std::fabs(0.1*HI(i,j,k)/HIdot[i]),
+                                    dt-ttot[i],
+                                    0.5*dt);
+
+      if (ddom[i] > 1.e8  && edot[i] > 0. &&
+          my_chemistry->primordial_chemistry > 1)  {
+        // here, we ensure that that the equilibrium mass density of
+        // Hydrogen changes by 10% or less
+        double Heq_div_dHeqdt = calc_Heq_div_dHeqdt_(
+          my_chemistry, my_rates, dlogtem, logTlininterp_buf,
+          kcr_buf.data[CollisionalRxnLUT::k13],
+          kcr_buf.data[CollisionalRxnLUT::k22],
+          d(i,j,k), tgas, p2d, edot, i
+        );
+
+        dtit[i] = std::fmin(dtit[i], 0.1*Heq_div_dHeqdt);
+      }
+
+      if (iter > 10) {
```

Comment:
I believe the reasoning for this predates me. I genuinely wonder why we would want to do this. If I understand correctly, we could temporarily find ourselves in a state with short subcycle timesteps, get through it, and then be forced to very slowly increase the timestep again.

I'll leave this comment only to keep a record since this isn't really the place to interrogate whether this should be here.